apps/dashboard/app/models/posix_file.rb

```diff
@@ -0,0 +1,176 @@
+    }
+  end
+
+  def directory?
+    path.stat.directory?
+  end
+
+  def ls
+    path.each_child.map do |child_path|
+      PosixFile.stat(child_path)
+    end.sort_by { |p| p[:directory] ? 0 : 1 }
+  end
+
+  def can_download_as_zip?(timeout: Configuration.file_download_dir_timeout, download_directory_size_limit: Configuration.file_download_dir_max)
+    can_download = false
+    error = nil
+
+    if ! (path.directory? && path.readable? && path.executable?)
+      error = I18n.t('dashboard.files_directory_download_unauthorized')
+    else
+      # Determine the size of the directory.
+      o, e, s = Open3.capture3("timeout", "#{timeout}s", "du", "-cbs", path.to_s)
+
+      # Catch SIGTERM.
+      if s.exitstatus == 124
+        error = I18n.t('dashboard.files_directory_size_calculation_timeout')
+      elsif ! s.success?
+        error = I18n.t('dashboard.files_directory_size_unknown', exit_code: s, error: e)
+      else
+        # Example output from: du -cbs $path
+        #
+        #    496184  .
+        #    64      ./ood-portal-generator/lib/ood_portal_generator
+        #    72      ./ood-portal-generator/lib
+        #    24      ./ood-portal-generator/templates
+        #    40      ./ood-portal-generator/share
+        #    576     ./ood-portal-generator
+        #
+        size = o&.split&.first
+
+        if size.blank?
+          error = I18n.t('dashboard.files_directory_size_calculation_error')
+        elsif size.to_i > download_directory_size_limit
+          error = I18n.t('dashboard.files_directory_too_large', download_directory_size_limit: download_directory_size_limit)
+        elsif size.to_i == 0
+          error = I18n.t('dashboard.files_directory_download_size_0', cmd: "timeout 10s du -cbs #{path.to_s}")
+        else
+          can_download = true
+        end
+      end
+    end
+
+    return can_download, error
+  end
+
+  def mime_type
+    type = %x[ file -b --mime-type #{path.to_s.shellescape} ].strip
+
+    # unfortunately, with the file command, we have this:
+    #
+    # > If the file named by the file operand does not exist, cannot be read,
+    # > or the type of the file named by the file operand cannot be determined,
+    # > this is not be considered an error that affects the exit status.
+    #
+    # so instead we validate it against a regex that match mimetypes
+    raise "not valid mimetype: #{type}" unless type =~ /^\w+\/[-+\.\w]+$/
+
+    # if you touch a file and it is empty, the mime type is "inode/x-empty"
+    # but in our interaction with the file we would treat this as "text/plain"
+    # so we return "text/plain" so web browsers treat it as so
+    if type == "inode/x-empty"
+      "text/plain"
+    else
+      type
+    end
+  end
+
+  def basename
+    path.basename
+  end
+
+  def descend
+    path.descend
+  end
+
+  def parent
+    path.parent
+  end
+
+  def join(*args)
+    path.join(*args)
+  end
+
+  def to_s
+    path.to_s
+  end
```

Comment:
You can just use Ruby's delegate directly. I'm suggesting this change but can you put it at the top near `attr_reader path`.

```suggestion
delegate :basename, :descend, :parent, :join, :to_s to: :path
```

Reply:
Fixed.